synapse/api/auth/msc3861_delegated.py

```diff
@@ -144,50 +153,75 @@ async def _introspect_token(self, token: str) -> IntrospectionToken:
-            headers=headers,
-        )
+        # check the cache before doing a request
+        introspection_token = self._token_cache.get(token, None)
+
+        expired = False
+        if introspection_token:
+            # check the expiration field of the token (if it exists)
+            exp = introspection_token.get("exp", None)
+            if exp:
+                time_now = self._clock.time_msec()
+                expired = time_now > exp
+
+        if not introspection_token or expired:
+            metadata = await self._issuer_metadata.get()
+            introspection_endpoint = metadata.get("introspection_endpoint")
+            raw_headers: Dict[str, str] = {
+                "Content-Type": "application/x-www-form-urlencoded",
+                "User-Agent": str(self._http_client.user_agent, "utf-8"),
+                "Accept": "application/json",
+            }
+
+            args = {"token": token, "token_type_hint": "access_token"}
+            body = urlencode(args, True)
+
+            # Fill the body/headers with credentials
+            uri, raw_headers, body = self._client_auth.prepare(
+                method="POST",
+                uri=introspection_endpoint,
+                headers=raw_headers,
+                body=body,
+            )
+            headers = Headers({k: [v] for (k, v) in raw_headers.items()})
+
+            # Do the actual request
+            # We're not using the SimpleHttpClient util methods as we don't want to
+            # check the HTTP status code, and we do the body encoding ourselves.
+            response = await self._http_client.request(
+                method="POST",
+                uri=uri,
+                data=body.encode("utf-8"),
+                headers=headers,
+            )
 
-        resp_body = await make_deferred_yieldable(readBody(response))
+            resp_body = await make_deferred_yieldable(readBody(response))
 
-        if response.code < 200 or response.code >= 300:
-            raise HttpResponseException(
-                response.code,
-                response.phrase.decode("ascii", errors="replace"),
-                resp_body,
-            )
+            if response.code < 200 or response.code >= 300:
+                raise HttpResponseException(
+                    response.code,
+                    response.phrase.decode("ascii", errors="replace"),
+                    resp_body,
+                )
 
-        resp = json_decoder.decode(resp_body.decode("utf-8"))
+            resp = json_decoder.decode(resp_body.decode("utf-8"))
 
-        if not isinstance(resp, dict):
-            raise ValueError(
-                "The introspection endpoint returned an invalid JSON response."
-            )
+            if not isinstance(resp, dict):
+                raise ValueError(
+                    "The introspection endpoint returned an invalid JSON response."
+                )
+
+            expiration = resp.get("exp", None)
+            if expiration:
```

Comment:
I am unsure if this is necessary - would the authorizing server ever send an introspection response with an expired `exp` field?

Reply:
I think its a worthwhile sanity check. It certainly could happen if there is *significant* clock drift between the two servers, but yeah it shouldn't happen in a healthy deployment